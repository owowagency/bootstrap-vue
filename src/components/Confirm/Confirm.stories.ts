import Btn from '@/components/Button';
import Confirm from '.';

export default {
    title: 'Components/Confirm',
    component: Confirm,
    argTypes: {},
};

const template = (args) => ({
    components: {Btn, Confirm},
    setup: () => ({args}),
    template: `
        <Btn
            data-bs-toggle="modal"
            data-bs-target="#confirm"
        >
            Launch demo confirm
        </Btn>

        <Confirm
            id="confirm"
            v-bind="args"
        />
    `,
});

export const Default = template.bind({});
