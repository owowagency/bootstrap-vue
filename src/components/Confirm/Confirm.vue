<template>
    <Modal
        ref="modal"
        :title="title"
    >
        <!-- @slot Displays content inside the confirm modal -->
        <slot>
            {{ body }}
        </slot>

        <template #footer>
            <!-- @slot Displays content inside the cancel button -->
            <slot name="buttonCancel">
                <Button
                    class="button-cancel"
                    :outline="cancelOutline"
                    :variant="cancelVariant"
                    @click="cancel"
                >
                    {{ cancelText }}
                </Button>
            </slot>

            <!-- @slot Displays content inside the ok button -->
            <slot name="buttonOk">
                <Button
                    class="button-ok"
                    :variant="okVariant"
                    @click="ok"
                >
                    {{ okText }}
                </Button>
            </slot>
        </template>
    </Modal>
</template>

<script lang="ts">
import {ComponentPublicInstance, ref} from 'vue';
import Button from '@/components/Button';
import Modal from '@/components/Modal';
import {variantProp} from '@/composables/useVariant';
</script>

<script lang="ts" setup>
defineProps({
    /**
     * The body of the confirm modal
     */
    body: {
        type: String,
        default: 'Are you sure?',
    },
    /**
     * Changes the cancel button to an outline variant
     */
    cancelOutline: {
        type: Boolean,
        default: false,
    },
    /**
     * The text of the cancel button
     */
    cancelText: {
        type: String,
        default: 'No',
    },
    /**
     * The variant of the cancel button
     */
    cancelVariant: variantProp('light'),
    /**
     * The text of the ok button
     */
    okText: {
        type: String,
        default: 'Yes',
    },
    /**
     * The variant of the ok button
     */
    okVariant: variantProp('primary'),
    /**
     * The title of the confirm modal
     */
    title: {
        type: String,
        default: 'Confirmation',
    },
});

const emit = defineEmits<{
    /**
     * Emitted when the cancel button is clicked
     */
    (e: 'cancelled'): void,
    /**
     * Emitted when the ok button is clicked
     */
    (e: 'confirmed'): void,
}>();

const cancel = () => {
    emit('cancelled');

    hide();
};

const ok = () => {
    emit('confirmed');

    hide();
};

const modal = ref<ComponentPublicInstance<typeof Modal>>();

const hide = () => modal.value.bsModal?.hide();

const show = () => modal.value.bsModal?.show();

const toggle = () => modal.value.bsModal?.toggle();

defineExpose({hide, show, toggle});
</script>

<docs>
```vue
<Confirm
    body="Are you sure you want to delete this?"
    :cancelOutline="true"
/>
```

## Imported Props

| Prop name     | Description                  | Type                                             | Values | Default   |
| ------------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| cancelVariant | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'light'   |
| okVariant     | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |
</docs>
