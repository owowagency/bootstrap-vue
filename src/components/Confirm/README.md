# Confirm

```vue
<Confirm body="Are you sure you want to delete this?" :cancelOutline="true" />
```

## Imported Props

| Prop name     | Description                  | Type                                             | Values | Default   |
| ------------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| cancelVariant | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'light'   |
| okVariant     | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |

## Props

| Prop name     | Description                                     | Type                   | Values | Default         |
| ------------- | ----------------------------------------------- | ---------------------- | ------ | --------------- |
| body          | The body of the confirm modal                   | string                 | -      | 'Are you sure?' |
| cancelOutline | Changes the cancel button to an outline variant | boolean                | -      | false           |
| cancelText    | The text of the cancel button                   | string                 | -      | 'No'            |
| cancelVariant | The variant of the cancel button                | variantProp('light')   | -      |                 |
| okText        | The text of the ok button                       | string                 | -      | 'Yes'           |
| okVariant     | The variant of the ok button                    | variantProp('primary') | -      |                 |
| title         | The title of the confirm modal                  | string                 | -      | 'Confirmation'  |

## Slots

| Name         | Description                               | Bindings |
| ------------ | ----------------------------------------- | -------- |
| default      | Displays content inside the confirm modal |          |
| buttonCancel | Displays content inside the cancel button |          |
| buttonOk     | Displays content inside the ok button     |          |

## Events

| Event name | Properties | Description                               |
| ---------- | ---------- | ----------------------------------------- |
| cancelled  |            | Emitted when the cancel button is clicked |
| confirmed  |            | Emitted when the ok button is clicked     |
