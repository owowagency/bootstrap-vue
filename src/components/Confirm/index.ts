import Confirm from './Confirm.vue';

export default Confirm;
