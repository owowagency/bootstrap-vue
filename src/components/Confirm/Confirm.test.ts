import {describe, expect, it} from 'vitest';
import Confirm from '.';
import {mount} from '@vue/test-utils';

describe('template', () => {
    componentRenderTest(
        Confirm,
        // Force an id to prevent snapshots from failing because of uuid.
        {props: {id: 'dont-change-please'}},
        false,
    );

    componentSlotRenderTest(Confirm);

    componentSlotRenderTest(Confirm, 'buttonCancel', {shallow: false});

    componentSlotRenderTest(Confirm, 'buttonOk', {shallow: false});

    const emits = [
        {el: '.button-cancel', emit: 'cancelled'},
        {el: '.button-ok', emit: 'confirmed'},
    ];

    for (const {el, emit} of emits) {
        it(`emits '${emit}' when ${el} is clicked`, async() => {
            const wrapper = await mount(Confirm);

            await wrapper.find(el).trigger('click');

            expect(wrapper.emitted()[emit]).toHaveLength(1);
        });
    }
});
