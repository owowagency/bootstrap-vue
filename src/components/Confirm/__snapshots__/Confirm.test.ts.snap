// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<div id=\\"dont-change-please\\" class=\\"modal\\">
  <div class=\\"modal-dialog modal-md\\">
    <div class=\\"modal-content\\">
      <div class=\\"modal-header\\">
        <!-- @slot Displays the header of the modal -->
        <h5 class=\\"modal-title\\">
          <!-- @slot Displays the header title of the modal -->Confirmation
        </h5><button class=\\"btn-close\\" data-bs-dismiss=\\"modal\\" type=\\"button\\"></button>
      </div>
      <div class=\\"modal-body\\">
        <!-- @slot Displays the body of the modal -->Are you sure?
      </div>
      <div class=\\"modal-footer\\">
        <!-- @slot Displays the footer of the modal -->
        <!-- @slot Displays content inside the cancel button --><button class=\\"btn btn-md btn-light button-cancel\\">
          <!-- @slot Displays content inside the button -->No
        </button><!-- @slot Displays content inside the ok button --><button class=\\"btn btn-md btn-primary button-ok\\">
          <!-- @slot Displays content inside the button -->Yes
        </button>
      </div>
    </div>
  </div>
</div>"
`;
